org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/IssueLocationsView.java

```diff
@@ -0,0 +1,402 @@
+        SonarLintLogger.get().error("Unable to read positions", e);
+        return Collections.emptyList();
+      }
+    }
+
+    @Override
+    public Object getParent(Object element) {
+      return null;
+    }
+
+    @Override
+    public boolean hasChildren(Object element) {
+      return getChildren(element).length > 0;
+    }
+
+    @Override
+    public void dispose() {
+      // Do nothing
+    }
+
+    @Override
+    public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
+      // Do nothing
+    }
+
+  }
+
+  private static class LocationsViewer extends TreeViewer {
+
+    public LocationsViewer(Tree tree) {
+      super(tree);
+      setAutoExpandLevel(ALL_LEVELS);
+      setUseHashlookup(true);
+    }
+  }
+
+  private static class LocationsTreeLabelProvider extends LabelProvider {
+
+    @Override
+    public String getText(Object element) {
+      if (element instanceof RootNode) {
+        return ((RootNode) element).getRoot().getAttribute(IMarker.MESSAGE, "No message");
+      } else if (element instanceof FlowRootNode) {
+        return ((FlowRootNode) element).getLabel();
+      } else if (element instanceof String) {
+        return (String) element;
+      }
+      return ((ExtraPosition) element).getMessage();
+    }
+
+    @Override
+    public Image getImage(Object element) {
+      if (element instanceof RootNode) {
+        return SonarLintImages.IMG_ISSUE;
+      }
+      return super.getImage(element);
+    }
+
+  }
+
+  public void setInput(@Nullable IMarker sonarlintMarker) {
+    if (sonarlintMarker == null) {
+      clearInput();
+      return;
+    }
+    // Find IFile and open Editor
+    // Super defensing programming because we don't really understand what is initialized at startup (SLE-122)
+    IWorkbenchWindow window = PlatformUI.getWorkbench().getActiveWorkbenchWindow();
+    if (window == null) {
+      showMarkerWithNoEditor(sonarlintMarker);
+      return;
+    }
+    IWorkbenchPage page = window.getActivePage();
+    if (page == null) {
+      showMarkerWithNoEditor(sonarlintMarker);
+      return;
+    }
+    for (IEditorReference editor : page.getEditorReferences()) {
+      IEditorInput editorInput;
+      try {
+        editorInput = editor.getEditorInput();
+      } catch (PartInitException e) {
+        SonarLintLogger.get().error("Unable to restore editor", e);
+        continue;
+      }
+      // Cast needed for older Eclipse versions
+      IFile file = (IFile) editorInput.getAdapter(IFile.class);
+      if (file != null && sonarlintMarker.getResource().equals(file)) {
+        IEditorPart editorPart = editor.getEditor(false);
+        if (editorPart instanceof ITextEditor) {
+          this.currentEditor = (ITextEditor) editorPart;
+          this.currentDoc = currentEditor.getDocumentProvider().getDocument(editorPart.getEditorInput());
+          this.currentFile = file;
+          locationsViewer.setInput(sonarlintMarker);
+          try {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)